File: pyTDA/Pila.py
from ObjNode import ObjNode as Node


class Pila:
    def __init__(self):
        self._inicio = None
        self._fin = None
        self._size = 0

    def apilar(self, valor):
        nuevo = Node(valor)
        if (self._inicio == None):
            self._inicio = self._fin = nuevo
            self._inicio.setSiguiente(self._fin)
            self._fin.setAnterior(self._inicio)
        else:
            self._fin.setSiguiente(nuevo)
            nuevo.setAnterior(self._fin)
            self._fin = nuevo
        self._size += 1

    def desapilar(self):
        if (self._inicio != None):
            if (self._fin == self._inicio):
                copia = self._fin.getValor()
                self._inicio = self._fin = None
                self._size -= 1
                return copia
            else:
                copia = self._fin.getValor()
                self._fin = self._fin.getAnterior()
                self._fin.setSiguiente(None)
                self._size -= 1
                return copia
        return None

    def tope(self):
        if (self._fin != None):
            return self._fin.getValor()
        return None

    def esVacia(self) -> bool:
        if (self._inicio == None):
            return True
        return False

    def size(self):
        return self._size

    def iter(self):
        actual = self._fin
        if (actual != None):
            while (True):
                yield actual.getValor()
                if (actual == self._inicio):
                    break
                actual = actual.getAnterior()

File: edd/Pila.js
"use strict";
class Nodo {
    constructor(valor) {
        this.valor = valor;
        this.arriba = null;
        this.abajo = null;
    }
}
class Pila {
    constructor() {
        this.fondo = null;
        this.tope = null;
        this.size = 0;
    }
    apilar(valor) {
        let nuevo = new Nodo(valor);
        if (this.tope == null) {
            this.tope = this.fondo = nuevo;
            this.tope.abajo = this.fondo;
            this.fondo.arriba = this.tope;
        } else {
            nuevo.abajo = this.tope;
            this.tope.arriba = nuevo;
            this.tope = nuevo;
        }
        this.size += 1;
    }
    desapilar() {
        if (this.tope != null) {
            const valu = this.tope.valor;
            this.tope = this.tope.abajo
            if (this.tope != null)
                this.tope.arriba = null;
            return valu;
        }
        return null;
    }
    *iter() {
        let actual = this.tope;
        while (actual != null) {
            yield actual.valor;
            actual = actual.abajo;
        }
    }

} export default Pila;